File: library/src/main/java/renetik/android/java/event/CSEvent.java
package renetik.android.java.event;

public interface CSEvent<T> {

    CSEventRegistration add(CSListener<T> listener);

    void fire(T arg);

    void clear();

    interface CSEventRegistration {
        void cancel();

        CSEvent<?> event();

        CSEventRegistration setActive(boolean active);

        boolean isActive();
    }
}

File: library/src/main/java/renetik/android/java/event/CSEvent.kt
package renetik.android.java.event

fun <T> fire(event: CSEvent<T>, argument: T) {
    event.fire(argument)
}

fun fire(eventVoid: CSEvent<Unit>) {
    eventVoid.fire(Unit)
}

fun <T> event(): CSEvent<T> {
    return CSEventImpl()
}

interface CSEvent<T> {

    fun run(listener: (registration: CSEventRegistration, argument: T) -> Unit): CSEventRegistration

    fun fire(argument: T)

    fun clear()

    interface CSEventRegistration {

        var isActive: Boolean

        fun cancel()

        fun event(): CSEvent<*>
    }
}

fun <T> CSEvent<T>.execute(function: (argument: T) -> Unit): CSEvent.CSEventRegistration {
    return this.run { _, argument -> function(argument) }
}